File: sphone-api/Domain/Products/Money.cs
using SPhone.Domain.Common;

namespace SPhone.Domain.Products;

public sealed class Money : ValueObject
{
    public decimal Amount { get; }
    public string Currency { get; }

    public Money(decimal amount, string currency = "THB")
    {
        if (amount < 0) throw new DomainException("Amount cannot be negative.");
        if (string.IsNullOrWhiteSpace(currency)) throw new DomainException("Currency is required.");
        Amount = amount;
        Currency = currency.ToUpperInvariant();
    }

    public Money Add(Money other)
    {
        if (Currency != other.Currency) throw new DomainException("Cannot add different currencies.");
        return new Money(Amount + other.Amount, Currency);
    }

    public Money Subtract(Money other)
    {
        if (Currency != other.Currency) throw new DomainException("Cannot subtract different currencies.");
        return new Money(Amount - other.Amount, Currency);
    }

    public Money Multiply(decimal factor) => new(Amount * factor, Currency);

    protected override IEnumerable<object?> GetEqualityComponents()
    {
        yield return Amount;
        yield return Currency;
    }

    public override string ToString() => $"{Amount:N2} {Currency}";
}

File: sphone-api/src/domain/products/Money.ts
import { ValueObject } from '../shared/ValueObject'

export class Money extends ValueObject<{ amount: number; currency: string }> {
  private constructor(amount: number, currency: string) {
    super({ amount, currency })
  }

  static create(amount: number, currency = 'THB'): Money {
    if (amount < 0) throw new Error('Amount cannot be negative')
    return new Money(amount, currency)
  }

  get amount(): number { return this._value.amount }
  get currency(): string { return this._value.currency }

  add(other: Money): Money {
    return new Money(this.amount + other.amount, this.currency)
  }

  multiply(factor: number): Money {
    return new Money(Math.round(this.amount * factor * 100) / 100, this.currency)
  }
}
